### PR TITLE
ARROW-3831: [C++] Add support for returning decompressed size

### PR DESCRIPTION
**Pull request by kou**

### CODE REVIEW

**Comment by wesm** [comment]
```
[ RUN      ] TestGZip/CodecTest.CodecRoundtrip/0
..\src\arrow\util\compression-test.cc(88): error:       Expected: data.size()
      Which is: 0
To be equal to: actual_decompressed_size
      Which is: -3689348814741910324
..\src\arrow\util\compression-test.cc(88): error:       Expected: data.size()
      Which is: 0
To be equal to: actual_decompressed_size
      Which is: -3689348814741910324
```

**Comment by kou** [comment]
I've fixed it.

**Review by wesm — APPROVED** [review]
+1. Thanks @kou! I opened https://issues.apache.org/jira/browse/ARROW-3877 about also adding support for decompression output size hints (which are provided by some of these libraries)

File: cpp/src/arrow/util/compression_snappy.cc [hunk 1]
```diff
@@ -42,8 +42,22 @@ Status SnappyCodec::MakeDecompressor(std::shared_ptr<Decompressor>* out) {
...
+                               int64_t ARROW_ARG_UNUSED(output_buffer_len),
+                               uint8_t* output_buffer, int64_t* output_len) {
+  if (output_len) {
+    size_t decompressed_size;
+    if (!snappy::GetUncompressedLength(reinterpret_cast<const char*>(input),
+                                       static_cast<size_t>(input_len),
+                                       &decompressed_size)) {
+      return Status::IOError("Corrupt snappy compressed data.");
+    }
+    *output_len = static_cast<int64_t>(decompressed_size);
```
Comment (pitrou) [inline cpp/src/arrow/util/compression_snappy.cc]: You don't check that `output_buffer_len` is large enough here?

File: cpp/src/arrow/util/compression-test.cc [hunk 2]
```diff
@@ -79,6 +79,14 @@ void CheckCodecRoundtrip(Compression::type ctype, const vector<uint8_t>& data) {
 
   ASSERT_EQ(data, decompressed);
 
+  // decompress with size with c2
```
Comment (pitrou) [inline cpp/src/arrow/util/compression-test.cc]: It would be useful to test what happens if the size is wrong...

**Comment by wesm** [comment]
I opened https://issues.apache.org/jira/browse/ARROW-3886 about adding another test

**Comment by kou** [comment]
I've created a pull request for it: #3057

**Pull request closed without merging**